cvmfs/export_plugin/fs_traversal.cc

```diff
@@ -0,0 +1,793 @@
+      atomic_inc64(&copy_queue);
+    } else {
+      if (!copyFile(src, src_ident, dest, dest_data, pstats)) {
+        LogCvmfs(kLogCvmfs, kLogStderr,
+          "Failed to copy %s->%s : %d : %s",
+          entry, dest_data, errno, strerror(errno));
+        errno = 0;
+        result = false;
+      }
+      free(src_ident);
+      pstats->Lookup(SHRINKWRAP_STAT_FILE_COUNT)->Inc();
+    }
+  } else {
+    pstats->Lookup(SHRINKWRAP_STAT_DEDUPED_FILES)->Inc();
+    pstats->Lookup(SHRINKWRAP_STAT_DEDUPED_BYTES)->Xadd(src_st->st_size);
+  }
+
+  // Should probably happen in the copy function as it is parallel
+  // Also this needs to be separate from copyFile, the target file
+  // could already exist and the link needs to be created anyway.
+  if (result && dest->do_link(dest->context_, entry, dest_data)) {
+    LogCvmfs(kLogCvmfs, kLogStderr,
+      "Failed to link %s->%s : %d : %s",
+      entry, dest_data, errno, strerror(errno));
+    errno = 0;
+    result = false;
+  }
+  free(dest_data);
+  return result;
+}
+
+bool handle_dir(
+  struct fs_traversal *src,
+  struct cvmfs_attr *src_st,
+  struct fs_traversal *dest,
+  struct cvmfs_attr *dest_st,
+  const char *entry
+) {
+  if (dest->do_mkdir(dest->context_, entry, src_st)) {
+    if (errno == EEXIST) {
+      errno = 0;
+      if (dest->set_meta(dest->context_, entry, src_st)) {
+        LogCvmfs(kLogCvmfs, kLogStderr,
+          "Traversal failed to set_meta %s", entry);
+        return false;
+      }
+    } else {
+      LogCvmfs(kLogCvmfs, kLogStderr,
+        "Traversal failed to mkdir %s : %d : %s",
+        entry, errno, strerror(errno));
+      return false;
+    }
+  }
+  return true;
+}
+
+void add_dir_for_sync(const char *dir, bool recursive) {
+  dirs_.push_back(new RecDir(strdup(dir), recursive));
+}
+
+bool Sync(
+  const char *dir,
+  struct fs_traversal *src,
+  struct fs_traversal *dest,
+  bool recursive,
+  perf::Statistics *pstats,
+  bool do_fsck) {
+  int cmp = 0;
+
+  char **src_dir = NULL;
+  size_t src_len = 0;
+  ssize_t src_iter = -1;
+  char * src_entry = NULL;
+  struct cvmfs_attr *src_st = cvmfs_attr_init();
+  list_src_dir(src, dir, &src_dir, &src_len);
+
+  char **dest_dir = NULL;
+  size_t dest_len = 0;
+  ssize_t dest_iter = -1;
+  char * dest_entry = NULL;
+  struct cvmfs_attr *dest_st = cvmfs_attr_init();
+  dest->list_dir(dest->context_, dir, &dest_dir, &dest_len);
+  qsort(dest_dir, dest_len, sizeof(char *),  strcmp_list);
+
+  // While both have defined members to compare.
+  while (1) {
+    if (cmp <= 0) {
+      getNext(src, dir, src_dir, &src_entry, &src_iter,
+            &src_st, true, true, pstats);
+    }
+
+    if (cmp >= 0) {
+      getNext(dest, dir, dest_dir, &dest_entry, &dest_iter,
+              &dest_st, do_fsck, false, pstats);
+    } else {
+      // A destination entry was added
+      pstats->Lookup(SHRINKWRAP_STAT_DEST_ENTRIES)->Inc();
+    }
+
+    if (!src_entry && !dest_entry) {
```

Comment:
This break case is separate from the comparisons. Though computationally the same, we may want to separate them so it is clear what they are meant to signify.